### ISSUE TITLE
Implement exception-free `cxx::Future` and `cxx::Promise`

### ISSUE THREAD

**Issue by mossmaurice** [eclipse-iceoryx/iceoryx#1390]
## Brief feature description

Automotive bindings like `ara::com` are using `std::Future` and `std::Promise` in their APIs. However, they throw exceptions which ins not in line with the iceoryx coding guidelines.

## Detailed information

* [ ] Anlayse feasablity of such an implementation
   * [ ] Check how API could be as close to the STL but with using `cxx::expected`
* [ ] Implement exception-free, stack-based `cxx::Future`
* [ ] Implement exception-free, stack-based `cxx::Promise`

**Comment by MatthiasKillat** [comment]
@mossmaurice See https://github.com/MatthiasKillat/iceoryx/tree/iox-futures-experimental for an incomplete approach how to implement `std::future` if this is the actual problem.

Now there are a lot of subtle races that need to be solved when attempting an implementation, especially if no dynamic memory is allowed.

Furthermore `ara::future` is maybe allowed to behave entirely differently or expected to perform more functionality, like communication with a server in another process (this is naturally not the case for `std::future`). See https://github.com/eclipse-iceoryx/iceoryx/discussions/1389#discussioncomment-2923045 for some additional explanation.

**Comment by MatthiasKillat** [comment]
@hemalbavishi This week I am almost completely unavailable. We can do this maybe on Friday this week or, preferable, sometime next week (e.g. Monday). If Monday or later I could prepare more information/code. 

We should use a zoom meeting then which I would announce in https://github.com/eclipse-iceoryx/iceoryx/wiki/Deep-Dive-Sessions. I would need some preferred time frame to set this up (e.g. Monday afternoon CEST).

**Comment by hemalbavishi** [comment]
@MatthiasKillat : It sounds good. Let's have a call next week to discuss more in detail. My colleague has integrated code provided by you but we would like to understand race conditions to make it better.  Let's discuss more in detail on Monday - Aug. 01

**Comment by MatthiasKillat** [comment]
@hemalbavishi I will set up a zoom meeting and ping you again. Note that the existing code was not error free at all. The plan was to improve it and thoroughly test it. But we can discuss this in the call.
.

**Comment by MatthiasKillat** [comment]
@hemalbavishi After some discussion with my colleagues we considered that it would be better to discuss this in the next iceoryx developer meetup on August 4th (i.e. next Thursday). Would this still work?

At least for me this would be easier to rethink the issue again and prepare something.

**Comment by hemalbavishi** [comment]
@MatthiasKillat : Okay. I will try to share some points with you before the call. It would be great if you can also do the same after your work over weekend.

**Comment by MatthiasKillat** [comment]
@hemalbavishi Yes, I can provide more information in the iceoryx developer meetup.

**Comment by SomnathHolkar** [comment]
@MatthiasKillat : Please update the ticket with open issues for future and promise implementation.

**Comment by mossmaurice** [comment]
@SomnathHolkar Just a short FYI: @MatthiasKillat is currently on vacation and comes back in CW35.

**Comment by MatthiasKillat** [comment]
@mossmaurice @SomnathHolkar I updated the issue with the use and misuse cases and implementation considerations. This reflects the current state of implementation, which likely has subtle races (which can be fixed once understood). I will improve the implementation once I have time but I cannot commit on this in the next week.

**Comment by hemalbavishi** [comment]
Hello Matthias,

We didn't get much time to work on that. Do you have any update from our last discussion? We want to take this forward and have std::promise and std::future heap free. 

Thank you!

**Comment by MatthiasKillat** [comment]
@hemalbavishi We did not work on it either. I will look into it in the coming days.
